File: apps/apple/CascadeShared/Sync/SyncApi.swift
import Foundation

/// Base URL of cascade-sync-server. Empty disables the sync feature.
enum SyncConfig {
    static let apiBase = "https://sync.cascade.stephens.page"
    static var available: Bool { !apiBase.isEmpty }
}

struct SyncError: Error {
    let status: Int
    let message: String
}

struct VerifyResponse: Decodable {
    let sessionToken: String
    let email: String
}

struct ListeningResponse: Decodable {
    let serverTotalMs: Int64
}

/// Typed async client for the sync endpoints (URLSession).
struct SyncApi {
    private let base = SyncConfig.apiBase
    private let session = URLSession.shared

    func requestLink(email: String) async throws {
        _ = try await send("POST", "/auth/request", body: ["email": email], token: nil)
    }

    func verify(token: String) async throws -> VerifyResponse {
        let data = try await send("POST", "/auth/verify", body: ["token": token], token: nil)
        return try JSONDecoder().decode(VerifyResponse.self, from: data)
    }

    func logout(token: String) async throws {
        _ = try await send("POST", "/auth/logout", body: nil, token: token)
    }

    func putListening(token: String, deviceId: String, deviceTotalMs: Int64) async throws -> ListeningResponse {
        let data = try await send(
            "PUT", "/listening",
            body: ["deviceId": deviceId, "deviceTotalMs": deviceTotalMs],
            token: token)
        return try JSONDecoder().decode(ListeningResponse.self, from: data)
    }

    func deleteListening(token: String) async throws {
        _ = try await send("DELETE", "/listening", body: nil, token: token)
    }

    func deleteAccount(token: String) async throws {
        _ = try await send("DELETE", "/account", body: nil, token: token)
    }

    private func send(_ method: String, _ path: String, body: [String: Any]?, token: String?) async throws -> Data {
        guard let url = URL(string: base + path) else {
            throw SyncError(status: 0, message: "bad url")
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = 15
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body { req.httpBody = try JSONSerialization.data(withJSONObject: body) }

        let (data, response) = try await session.data(for: req)
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200...299).contains(code) else {
            throw SyncError(status: code, message: String(data: data, encoding: .utf8) ?? "request failed (\(code))")
        }
        return data
    }
}

File: apps/windows/Cascade/Services/SyncApi.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;

namespace Cascade.Services;

/// <summary>Base URL of cascade-sync-server. Empty disables the sync feature.</summary>
public static class SyncConfig
{
    public const string ApiBase = "https://sync.cascade.stephens.page";
    public static bool Available => !string.IsNullOrEmpty(ApiBase);
}

public sealed class SyncHttpException : Exception
{
    public int Status { get; }
    public SyncHttpException(int status, string message) : base(message) => Status = status;
}

public sealed record VerifyResponse(string SessionToken, string Email);
public sealed record ListeningResponse(long ServerTotalMs, long SyncedThroughMs);

/// <summary>Typed client for the sync endpoints (System.Net.Http + System.Text.Json).</summary>
public sealed class SyncApi
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(15) };

    public Task RequestLinkAsync(string email) =>
        // The "windows" hint makes the emailed link carry &app=windows, so the
        // web /auth page hands the token to this app via cascade:// rather than
        // consuming it in the browser.
        PostAsync("/auth/request", new { email, platform = "windows" }, null);

    public async Task<VerifyResponse> VerifyAsync(string token)
    {
        using var resp = await SendAsync(HttpMethod.Post, "/auth/verify", new { token }, null);
        return (await resp.Content.ReadFromJsonAsync<VerifyResponse>())!;
    }

    public Task LogoutAsync(string sessionToken) =>
        PostAsync("/auth/logout", null, sessionToken);

    public async Task<ListeningResponse> PutListeningAsync(string sessionToken, string deviceId, long deviceTotalMs)
    {
        using var resp = await SendAsync(HttpMethod.Put, "/listening",
            new { deviceId, deviceTotalMs }, sessionToken);
        return (await resp.Content.ReadFromJsonAsync<ListeningResponse>())!;
    }

    public Task DeleteListeningAsync(string sessionToken) =>
        SendVoidAsync(HttpMethod.Delete, "/listening", sessionToken);

    public Task DeleteAccountAsync(string sessionToken) =>
        SendVoidAsync(HttpMethod.Delete, "/account", sessionToken);

    private async Task PostAsync(string path, object? body, string? token)
    {
        using var _ = await SendAsync(HttpMethod.Post, path, body, token);
    }

    private async Task SendVoidAsync(HttpMethod method, string path, string? token)
    {
        using var _ = await SendAsync(method, path, null, token);
    }

    private static async Task<HttpResponseMessage> SendAsync(
        HttpMethod method, string path, object? body, string? token)
    {
        using var req = new HttpRequestMessage(method, $"{SyncConfig.ApiBase}{path}");
        if (token is not null)
            req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
        if (body is not null)
            req.Content = JsonContent.Create(body);
        var resp = await Http.SendAsync(req);
        if (!resp.IsSuccessStatusCode)
        {
            var text = await resp.Content.ReadAsStringAsync();
            resp.Dispose();
            throw new SyncHttpException((int)resp.StatusCode,
                string.IsNullOrEmpty(text) ? $"request failed ({(int)resp.StatusCode})" : text);
        }
        return resp;
    }
}
